subsys/net/lib/sockets/sockets_tls.c

```diff
@@ -2645,8 +4064,244 @@ static ssize_t recv_tls(struct tls_context *ctx, void *buf,
 
 	return recv_len;
 }
+#endif /* CONFIG_MBEDTLS */
 
 #if defined(CONFIG_NET_SOCKETS_ENABLE_DTLS)
+#if defined(CONFIG_WOLFSSL)
+static ssize_t recvfrom_dtls_common_wolfssl(struct tls_context *ctx, void *buf,
+                    size_t max_len, int flags,
+                    struct sockaddr *src_addr,
+                    socklen_t *addrlen, int *werr)
+{
+    int ret;
+    bool is_block = is_blocking(ctx->sock, flags);
+    k_timeout_t timeout;
+    k_timepoint_t end;
+
+    if (ctx->error != 0) {
+        errno = ctx->error;
+        return -1;
+    }
+
+    if (!is_block) {
+        timeout = K_NO_WAIT;
+    } else {
+        timeout = ctx->options.timeout_rx;
+    }
+
+    end = sys_timepoint_calc(timeout);
+
+    do {
+        size_t remaining;
+        *werr = 0;
+
+        ret = wolfSSL_read(ctx->wssl, buf, max_len);
+        if (ret < 0) {
+            *werr = wolfSSL_get_error(ctx->wssl, ret);
+            if (*werr == WOLFSSL_ERROR_WANT_READ ||
+                *werr == WOLFSSL_ERROR_WANT_WRITE) {
+                int timeout_dtls, timeout_sock, timeout_ms;
+
+                if (!is_block) {
+                    return ret;
+                }
+
+                /* Blocking timeout. */
+                timeout = sys_timepoint_timeout(end);
+                if (K_TIMEOUT_EQ(timeout, K_NO_WAIT)) {
+                    return ret;
+                }
+
+                timeout_dtls = wolfSSL_dtls_get_current_timeout(ctx->wssl);
+                timeout_sock = timeout_to_ms(&timeout);
+                if (timeout_sock == SYS_FOREVER_MS) {
+                    timeout_ms = MAX(timeout_dtls, timeout_sock);
+                } else {
+                    timeout_ms = MIN(timeout_dtls, timeout_sock);
+                }
+
+                /* Block. */
+                k_mutex_unlock(ctx->lock);
+                ret = wait_for_reason(ctx->sock, timeout_ms, *werr);
+                k_mutex_lock(ctx->lock, K_FOREVER);
+
+                if (ret == 0) {
+                    /* Retry. */
+                    continue;
+                } else {
+                    return -1;
+                }
+            } else {
+                return ret;
+            }
+        }
+
+        if (src_addr && addrlen) {
+            dtls_peer_address_get(ctx, src_addr, addrlen);
+        }
+
+        /* mbedtls_ssl_get_bytes_avail() indicate the data length
+         * remaining in the current datagram.
+         */
+        remaining = wolfSSL_pending(ctx->wssl);
+
+        /* No more data in the datagram, or dummy read. */
+        if ((remaining == 0) || (max_len == 0)) {
+            return ret;
+        }
+
+        if (flags & ZSOCK_MSG_TRUNC) {
+            ret += remaining;
+        }
+
+        for (int i = 0; i < remaining; i++) {
+            uint8_t b;
```

Comment:
Maybe it makes sense to be more efficient with the flushing?
```suggestion
            uint8_t b[10];
```

Reply:
A lot of this is just copy pasted from the original mbedtls is all, I agree and will make the change to flush more efficiently.